### PR TITLE
Changes Amith211 for BFA

### PR DESCRIPTION
**Pull request by coenvdwel**

### CODE REVIEW

File: Rewatch.lua [hunk 1]
```diff
@@ -7,7 +7,7 @@
 -- Please give full credit when you want to redistribute or modify this addon!
 
 
-local rewatch_versioni = 60007;
+local rewatch_versioni = 80000;
```
Comment (coenvdwel) [inline Rewatch.lua]: There is no need to keep Rewatch version aligned with WoW version :-D you just released 2 complete rework versions of Rewatch (7.0 and 8.0) 😀 🤣

Comment (coenvdwel) [reply Rewatch.lua]: (I choose to release 61000 for the BFA + Shaman support release)

File: Rewatch.lua [hunk 2]
```diff
@@ -92,6 +92,11 @@ function rewatch_OnLoad()
 				rewatch_load["ButtonSpells"] = { rewatch_loc["swiftmend"], rewatch_loc["naturescure"], rewatch_loc["ironbark"], rewatch_loc["healingtouch"], rewatch_loc["mushroom"] };
 				if(rewatch_load["Layout"] == "vertical") then rewatch_load["FrameColumns"] = 1; else rewatch_load["FrameColumns"] = 0; end;
 			end;
+
+			if(rewatch_version < 80000) then
+				rewatch_load["ButtonSpells"] = { rewatch_loc["swiftmend"], rewatch_loc["naturescure"], rewatch_loc["ironbark"], rewatch_loc["healingtouch"], rewatch_loc["mushroom"] };
+				if(rewatch_load["Layout"] == "vertical") then rewatch_load["FrameColumns"] = 1; else rewatch_load["FrameColumns"] = 0; end;
```
Comment (coenvdwel) [inline Rewatch.lua]: There's no need, this is the default from the previous version!

File: Rewatch.lua [hunk 3]
```diff
@@ -307,22 +312,38 @@ function rewatch_SetTooltip(data)
...
+	return spellName;
+end
+
 -- gets the spell ID of the highest rank of the specified spell
 -- spellName: the name of the spell to get the highest ranked spellId from
 -- return: the corresponding spellId
 function rewatch_GetSpellId(spellName)
 	-- get spell info and highest rank, return if the user can't cast it (not learned, etc)
-	local name, rank, icon = GetSpellInfo(spellName);
+	local name, rank, icon, _, _, _, spellId = GetSpellInfo(spellName);
```
Comment (coenvdwel) [inline Rewatch.lua]: My previous reviewer for BFA did not notice this is broken; are you testing this on a beta WoW version or something?

File: Rewatch.lua [hunk 4]
```diff
@@ -778,7 +799,7 @@ function rewatch_UpdateBar(spellName, player, stacks)
 	if(rewatch_bars[playerId][spellName]) then
 	
 		-- get buff duration
-		local a = select(7, UnitBuff(player, spellName, nil, "PLAYER"));
+		local a = select(6, AuraUtil.FindAuraByName(spellName, player, "PLAYER"));
```
Comment (coenvdwel) [inline Rewatch.lua]: Oeh, what is this AuraUtil? The patch I have for this change is looping over `UnitBuff(player, counter, filter)` where counter is incremented from 1 to 40, your solution seems more efficient? Can you clarify?

File: Rewatch.lua [hunk 5]
```diff
@@ -1403,130 +1424,142 @@ rewatch_events:SetScript("OnEvent", function(timestamp, event, unitGUID, effect,
...
-				val = rewatch_GetPowerBarColor("RAGE");
-				rewatch_bars[playerId]["ManaBar"]:SetStatusBarColor(val.r, val.g, val.b, 1);
-			end;
-		-- else, if it was Clearcasting being applied to us
-		elseif((spell == rewatch_loc["clearcasting"]) and (targetName == UnitName("player"))) then
-			for n=1,rewatch_i-1 do val = rewatch_bars[n]; if(val) then
-				if(val[rewatch_loc["regrowth"]]) then
-					val[rewatch_loc["regrowth"].."Bar"]:SetStatusBarColor(1, 1, 1, 1);
+	-- buff applied/refreshed 
+	elseif(event == "COMBAT_LOG_EVENT_UNFILTERED") then
```
Comment (coenvdwel) [inline Rewatch.lua]: I'm not really sure what's happening here, the GitHub changelog has some issues with the spacing/lining that has changed. Another patch I've seen has a shorter solution, however, I cannot judge if that captures all your changes below. What are your thoughts?

https://github.com/coenvdwel/Rewatch/blob/61000/Rewatch.lua#L1450

**Comment by coenvdwel** [comment]
@Louis

File: Rewatch.lua (hunk 1 above)
Comment (Amith211) [reply Rewatch.lua]: Woops xD. I'll take a look at the 6100 release.

File: Rewatch.lua (hunk 2 above)
Comment (Amith211) [reply Rewatch.lua]: Doh, my bad!

File: Rewatch.lua (hunk 4 above)
Comment (Amith211) [reply Rewatch.lua]: It's a new Bfa function found in FrameXML/AuraUtil.lua. It 'Finds the first aura that matches the name'. 
There are some important notes to bear in mind:
* '[A]ura names are not unique!'
* '[A]ura names are localized, what works in one locale might not work in another consider that in English two auras might have different names, but once localized they have the same name, so even using the localized aura name in a search it could result in different behaviour'.
* '[T]he unit could have multiple auras with the same name, this will only find the first'.

As this is used only for class spells, if I'm not mistaken, I don't believe we are likely to run into any problems.

See [WowPedia](https://wow.gamepedia.com/API_AuraUtil.FindAuraByName).

File: Rewatch.lua (hunk 5 above)
Comment (Amith211) [reply Rewatch.lua]: I'm glad I'm not the only one who had difficulties following using haha.

Blizzard have made some changes with the COMBAT_LOG_EVENT in that it no longer has any payload. CombatLogGetCurrentEventInfo() is to be used to retrieve that.

The principal difference between my changes and the one mentioned is that, from what I have observed in game, the if statements using the 'effect' variable are checking the combat log and as such are only (at least the ones we are checking for) present when COMBAT_LOG_EVENT is raised, hence we only need to make those checks if the event in question is a COMBAT_LOG_EVENT.

What I have done is wrap those elseif statements in the elseif you have referenced above.

I accept that my implementation when setting the variables might not be as elegant as the other solution, I make no comment on that point and leave that to your best judgement.

To summarise, this change is made to prevent unnecessary checks when an event is raised that is not a COMBAT_LOG_EVENT.

See [8.0 Release Notes for Wow Addon Developers - Blue post](https://us.battle.net/forums/en/wow/topic/20762318007#post-1)

File: Rewatch.lua [hunk 6]
```diff
@@ -307,24 +312,58 @@ function rewatch_SetTooltip(data)
...
+		repeat
+			if ((ispell == name) and ((rank == irank) or (not irank) or (rank == nil or irank == nil))) then return i; end;
+			i, ispell, irank = i+1, GetSpellBookItemName(i+1, BOOKTYPE_SPELL);
+		until (not ispell);
+		
+		-- return default -1
+		return -1;
+	end);
+
+	return ispellId;
```
Comment (Amith211) [inline Rewatch.lua]: Blizzard have made some major back end changes with the Spell System API, with a focus on SpellId. The spell text fields, except SpellName are now loaded on demand.
This means that 'If you call these functions, the listed return value may be nil or empty, since the data is not available at the time of the function call.'

GetSpellBookItemName -- # 2 (spell name subtext) is affected by this change. (I believe this is retrieved by irank).

I have made the above changes folling the Blizzard examples.

See [ 8.0 Release Notes for Wow Addon Developers - Blue post](https://us.battle.net/forums/en/wow/topic/20762318007#post-1)

Comment (Amith211) [reply Rewatch.lua]: Apologies , I wasn't sure how many lines it would include. I refer to lines 337-352.

File: Rewatch.lua (hunk 3 above)
Comment (Amith211) [reply Rewatch.lua]: This is not broken per se, but it is required for the changes at lines 338 ff.

I refer to my comment, mistakenly, added at line 352.

Comment (coenvdwel) [reply Rewatch.lua]: Sorry, I fail to understand what this change would actually mean. I mean, I can see what it does, but I do not understand if or how it is better. Probably this increases performance, since you only loop the ranks of that one spell instead of the full spell book, but I cannot see the 'cost' of the `Spell:CreateFromSpellID` and `spell:ContinueOnSpellLoad`.

I will follow your judgement here; please tell me if you think this should be copied to the 61000 release version, and I will! 😁

Comment (Amith211) [reply Rewatch.lua]: I'm not sure if there are any performance improvements with these changes, but I believe `spell:ContinueOnSpellLoad` is necessary to ensure that certain spell information is returned (rather than nil or empty) after the Blizzard 'Load on demand' backend  changes.

On that basis, I think these changes need to be included in the 61000 version.
